### DISCUSSION TITLE
Is it possible for `worker.stop()` to return a promise?

### DISCUSSION THREAD

**Discussion by bertdeblock** [mswjs/msw#1092] [Ideas]
I _think_ `worker.stop()` used to return a promise in older versions of MSW? This doesn't seem to be the case anymore.
Is it possible to bring back this functionality or is there a technical reason that prevents `worker.stop()` from returning a promise?
The reason I'm asking is because, we have a scenario where we call `worker.stop()` and immediately do a `fetch` call afterwards. Sometimes the call works as intended, but other times this leaves the `fetch` promise in an infinite pending state. The worker seems to intercept the fetch call, but is not able to send back the original response because it has also received the stop command (or something to that extent).

Please let me know if you need any more info etc.

Thanks for the great library BTW, it has been very useful to us 👍!

**Comment by kettanaito** [comment]
Hey, @bertdeblock. Thanks for your kind words!

I think `.stop()` used to return a Promise a very long time ago because it unregistered the worker. That's no longer the case, and now calling `.stop()` only sends a special event to the worker saying that it should ignore any requests coming from this particular client:

https://github.com/mswjs/msw/blob/6e7ec2871f5f82fbf228b803df565286b1d4dbfc/src/setupWorker/stop/createStop.ts#L13

On the worker's side, this means removing the client's ID from the internal set of active client IDs:

https://github.com/mswjs/msw/blob/6e7ec2871f5f82fbf228b803df565286b1d4dbfc/src/mockServiceWorker.js#L64-L67

Once removed, all requests from this client _will still be intercepted_ but not mocked. Instead, they will be performed as-is:

https://github.com/mswjs/msw/blob/6e7ec2871f5f82fbf228b803df565286b1d4dbfc/src/mockServiceWorker.js#L151-L157

Do you mind setting a few breakpoints in the `mockServiceWorker.js` in your browser's DevTools and seeing what exactly happens when your request gets stuck? 

Well, I think it's possible that a request happens, triggers the `fetch` listener, begins processing, and meanwhile `.stop()` is called, removing the client's ID. I'm still curious when exactly this hangs and why.

I guess we can consider making the deactivation events bi-directional so that the worker would send back something like `DEACTIVATE_SUCCESS` and the client will establish a Promise in `.stop()` awaiting that message.

**Comment by bertdeblock** [comment]
@kettanaito Thanks for the response. I will take some time to reproduce the issue and provide you with more detailed info. Won't be right away, but I'll make it happen!

**Reply by kettanaito** [reply]
Take your time. I'm looking forward to diving into this issue together. At this point, it's about learning as much as I can about the use case and providing you with proper guidance on how to achieve it.

**Comment by bertdeblock** [comment]
FWIW, this has become less of an issue for us.
We now have more control over the `fetch` call mentioned in the initial post.
Meaning, we can properly await the call and invoke `worker.stop()` afterwards.
Feel free to close this discussion if you see fit @kettanaito.
Thanks for the help!
